### PR TITLE
`39` semantic searching

### PR DESCRIPTION
**Pull request by AardhynLavender**
## Semantic Searching

Researching and implementing a prototype for semantic searching on articles

[`https://www.sbert.net/`](https://www.sbert.net/)

> The `Docker` environment takes a good **6-10 minutes** to build, and has issues communicating with the server running on the host machine. Nothing `Docker Compose` can't solve later...

## Todo

- [x] Flask application to allow the server to communicate with the search engine
- [x] ~~Store tensor of article names ( and other fields maybe ) in the database on creation.~~
- [x] ~~Route to retrieve article name vectors with keys ( Don't think a new table is needed for this )~~
- [x] `article.search.domain` route ( article `name` and `id` )
- [x] `article.search` ( article `name` and `id` )
- [x] ~~Search caching?~~

## Execution

Build the database, start the server, and seed the database from the scraper.

Configure `SEARCH_ENGINE_HOST` and `SEARCH_ENGINE_PORT` in `/server/.env`. Use an unused port of your choice and *localhost* for the host.

Configure these same variables in `/search/.env`.

Run the search engine with 

```plaintext
init.local.sh
```

Send your queries to the server with

```plaintext
/article.search?query=your search term here
```

### CODE REVIEW

**Review by gpseal — CHANGES_REQUESTED** [review]
Wow, a great addition to the app, really great work here. 
I had a go with a couple of queries, and it works really well.  Some results seem a little less related than others, but the first few seem to be spot on, I guess that's the nature of search.

Error handling is good as always, which is appreciated (was useful in testing)

Just a couple of minor things I have pointed out to fix, not sure if the first one is necessary, but take a look.

File: server/template.env [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
-DB_PASSWORD = 
\ No newline at end of file
+DB_HOST= 
+DB_PORT= 
+DB_NAME= 
+DB_USER=
+DB_PASSWORD= 
+
+# search engine
+SEARCH_ENGINE_PORT=
```
[the remark this answers is not in the document]
Comment (AardhynLavender) [reply server/template.env]: added 👍

File: search/init.local.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
+if [ -z .env ] ; then
+  echo "Unable to initalize search engine!"
+  echo "Missing `.env` file ❌";
+  exit 1;
+fi
+source .env;
+
+pip install -r requirements.txt
```
[the remark this answers is not in the document]
Comment (AardhynLavender) [reply search/init.local.sh:8]: I didn't have this problem on my machine... [seems to make sense though](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-user).

**Review by ArtemKech — APPROVED** [review]
Works great! It is a really cool feature. Works better than most of the searches on other websites :). I agree with Greg, that some of the results are quite irrelevant to what you are trying to search but the first few seem to be accurate. Good job!

**Pull request merged**